### PR TITLE
[DSS] DDP-7931: User registration endpoint

### PR DESCRIPTION
**Pull request by Trimax**
This PR contains a couple of things:
1. Implementation of the new endpoint for creation the user
2. Some code cleanup (unused methods and constants were removed)
3. UserProfile class builder class was replaced with Lombok's implementation

### CODE REVIEW

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+                throw new RuntimeException("Tenant wasn't found for study + " + payload.getStudyGuid());
+            }
+
+            final var centerDto = handle.attach(CenterProfileDao.class).findById(payload.getCenterId());
+            if (centerDto.isEmpty()) {
+                throw new RuntimeException("Center " + payload.getCenterId() + " does not exist");
+            }
+
+            final var clientId = handle.attach(ClientDao.class).getClientDao().insertClient(
+                    "DON'T KNOW", "DON'T KNOW", tenantDto.getId(), "DON'T KNOW");
```
Comment (Trimax) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java]: How exactly should I create the user without knowing all this auth0 stuff? Should it be sent together with payload from the client?

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+            final var user = handle.attach(UserDao.class).createUser(clientId, "DON'T KNOW");
```
Comment (Trimax) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java]: How exactly should I create the user without knowing all this auth0 stuff? Should it be sent together with payload from the client?

File: pepper-apis/dss-database/src/main/resources/db-changes/schema/DDP-7931.xml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+            <column name="center_study_id" type="bigint">
+                <constraints primaryKey="true" primaryKeyName="center_study_pk"/>
+            </column>
+            <column name="center_id" type="bigint">
+                <constraints references="center_profile(center_id)" foreignKeyName="center_study_center_fk"/>
+            </column>
+            <column name="study_id" type="bigint">
+                <constraints references="umbrella_study(umbrella_study_id)" foreignKeyName="center_study_study_fk"/>
+            </column>
+        </createTable>
```
Comment (Trimax) [inline pepper-apis/dss-database/src/main/resources/db-changes/schema/DDP-7931.xml:26]: This table is required in case the creation flow will be different. In case we will use some operator id instead of center id.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/CreateTemporaryUserRoute.java [hunk 4]
```diff
@@ -77,9 +77,10 @@ public Object handle(Request request, Response response, CreateTemporaryUserPayl
             try {
                 User user = handle.attach(UserDao.class).createTempUser(clientDto.getId());
                 //create profile
-                UserProfileDao profileDao = handle.attach(UserProfileDao.class);
-                profileDao.createProfile(new UserProfile.Builder(user.getId())
-                        .setPreferredLangId(languageId)
+                handle.attach(UserProfileDao.class).createProfile(UserProfile.builder()
+                        .userId(user.getId())
+                        .preferredLangId(languageId)
+                        .preferredLangCode(null)
```
Comment (bskinner) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/CreateTemporaryUserRoute.java:83]: We should be setting one of `preferredLangId` or `preferredLangCode`, but not both (even though it is `null`)- is there a way we can avoid this sort of duplicated call?

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java (hunk 1 above)
Comment (bskinner) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java]: We shouldn't be doing anything with the Auth0 client here. Setting up Auth0 clients is done during study initialization and, regardless, any users created via this API will not be bound to an Auth0 account.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java (hunk 2 above)
Comment (bskinner) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java]: Users created for a Center do not have an associated Auth0 account.

File: pepper-apis/dss-database/src/main/resources/changelog-master.xml [hunk 5]
```diff
@@ -2,7 +2,7 @@
 <databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
                    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                    xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
-                   http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.5.xsd">
+                   https://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.5.xsd">
```
Comment (bskinner) [inline pepper-apis/dss-database/src/main/resources/changelog-master.xml:5]: Why is this scheme being changed to `https`, but none of the others are?

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java [hunk 6]
```diff
@@ -368,6 +367,7 @@ private static void start() throws MalformedURLException {
         }
 
         post(API.REGISTRATION, new UserRegistrationRoute(interpreter, new TaskPubSubPublisher()), responseSerializer);
+        post(API.USER_SPECIFIC, new UserCreationRoute(), responseSerializer);
```
Comment (bskinner) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java]: If your intention is to add a new post endpoint at `/user`, I don't think this is going to do what you want it do to. The `USER_SPECIFIC` includes a `:userGuid` path parameter and, if this is creating a user, we do not have a GUID yet.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/json/UserCreationPayload.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    private String lastName;
+
+    @SerializedName("birthDate")
+    private LocalDate birthDate;
+
+    @SerializedName("sex")
+    private UserProfile.SexType sex;
+
+    @SerializedName("centerId")
+    private Long centerId;
```
Comment (bskinner) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/json/UserCreationPayload.java]: Why is `centerId` a `Long` here? Typically, all external identifiers (such as user guids, activity codes, instance identifiers, etc) are all opaque strings.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/db/dao/CenterProfileDao.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+import org.jdbi.v3.stringtemplate4.UseStringTemplateSqlLocator;
+
+import java.util.Optional;
+
+public interface CenterProfileDao extends SqlObject {
+    @UseStringTemplateSqlLocator
+    @SqlQuery("SELECT * FROM `center_profile` WHERE center_id = :id")
+    @RegisterConstructorMapper(CenterProfileDto.class)
+    Optional<CenterProfileDto> findById(@Bind("id") final Long id);
+}
```
Comment (bskinner) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/db/dao/CenterProfileDao.java:17]: The Center-related code is out-of-scope for this ticket since none of the management or study-related configuration code is in place. Leave this for future expansion, but it's not currently used.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/util/JavaHeapDumper.java [hunk 9]
```diff
@@ -1,60 +1,161 @@
...
-import com.google.cloud.storage.Blob;
-import com.google.cloud.storage.Storage;
-import com.sun.management.HotSpotDiagnosticMXBean;
-import lombok.extern.slf4j.Slf4j;
 import org.broadinstitute.ddp.exception.DDPException;
 
+import lombok.extern.slf4j.Slf4j;
+
 @Slf4j
 public class JavaHeapDumper {
```
Comment (bskinner) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/util/JavaHeapDumper.java:20]: This class was modified in order to fix compilation issues when using non-HotSpot JVMs (this was preventing my builds). This method isn't called during normal operation, and only called manually when debugging server memory issues so this shouldn't have a significant impact on the overall operation of the backend.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+    @Override
+    protected int getValidationErrorStatus() {
+        return HttpStatus.SC_UNPROCESSABLE_ENTITY;
+    }
+
+    @Override
+    public Object handle(Request request, Response response, UserCreationPayload payload) throws Exception {
+        var auth = RouteUtil.getDDPAuth(request);
+
+        if (false && auth.isAuthenticated() == false) {
```
Comment (bskinner) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java]: Known `if` bypass here- currently working on testing out the authentication issues. Expect this line to be updated once I get I reliable source of DSM tokens.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+            var error = new ApiError(ErrorCodes.AUTH_CANNOT_BE_DETERMINED, "Authentication required.");
+            throw ResponseUtil.haltError(HttpStatus.SC_FORBIDDEN, error);
+        } else {
+            /* LIE */
+            auth = new DDPAuth("https://domain.example", "abcTESTCLIENTdef", "OPERATOR", "totallyAValidJWT", null, "en");
+        }
```
Comment (bskinner) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java]: Another block of code that is going away once testing can be performed using DSM tokens.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+        final var operatorId = auth.getOperator();
+        final var domain = auth.getIssuer();
+        final var studyGuid = payload.getStudyGuid();
+
+        // Not sure if this is the proper check, but this is
+        // just to add some sort of safeguard until the filters are
+        // configured properly.
+        //
+        // Could this be replaced with an "isDSM" check?
+        if (false && auth.hasAdminAccessToStudy(studyGuid) == false) {
```
Comment (bskinner) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java]: Another block of code that is going away once testing can be performed using DSM tokens.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+            // needed below to set the `created_by_client_id` field in the user.
+            final var internalClientId = handle.attach(JdbiClient.class)
+                    .getClientIdByAuth0ClientAndDomain(requestorClientId, domain);
+            if (internalClientId.isEmpty()) {
+                /* Disabled for testing!
+                var error = new ApiError(ErrorCodes.NOT_FOUND,
+                        String.format("Auth0 client '%s' is not authorized for '%s'.",
+                            requestorClientId, domain));
+                throw ResponseUtil.haltError(HttpStatus.SC_UNAUTHORIZED, error);
+                */
```
Comment (bskinner) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java]: Needs to be re-enabled after testing can be performed with real tokens.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/db/dao/UserProfileSql.java [hunk 14]
```diff
@@ -19,7 +19,7 @@ public interface UserProfileSql extends SqlObject {
 
     @GetGeneratedKeys
     @SqlUpdate("insert into user_profile (user_id, first_name, last_name, sex, birth_date,"
-            + "        preferred_language_id, time_zone, do_not_contact, is_deceased, skip_language_popup)"
+            + "        preferred_language_id, time_zone, do_not_contact, is_deceased, skip_language_popup, email)"
             + " values (:userId, :firstName, :lastName, :sex, :birthDate,"
             + " :langId, :tz, :doNotContact, :isDeceased, :skipLanguagePopup)")
```
Comment (Trimax) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/db/dao/UserProfileSql.java]: An e-mail was added to the columns list, but the value seems to be missing

Comment (Trimax) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/db/dao/UserProfileSql.java]: In the list of arguments as well

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/db/dto/UserDto.java [hunk 15]
```diff
@@ -37,6 +41,14 @@ public class UserDto {
     @ColumnName("expires_at")
     Long expiresAtMillis;
 
+    public Optional<String> getAuth0UserId() {
+        return Optional.ofNullable(auth0UserId);
+    }
+
+    public Optional<String> getEmail() {
+        return Optional.ofNullable(email);
+    }
```
Comment (Trimax) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/db/dto/UserDto.java:50]: Do you think it is useful? I see that this method is being used only in one place - UserCreationRoute class.

The construction
```
model.getEmail().orElse(null);
```
looks a bit less readable for me that 
```
Optional.ofNullable(model.getEmail()).orElse(null);
```
or simple ternary operator
```
model.getEmail() == null ? null : model.getEmail();
```

Anyway, it's just a matter of taste, so if you believe returning optional is better, then let's leave it as is.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/json/UserCreationResponse.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+    @SerializedName("profile")
+    private Profile profile;
+
+    @SerializedName("center")
+    private String center;
+    
+    public UserCreationResponse(User user, UserProfile profile) {
+        this.guid = user.getGuid();
+        this.hruid = user.getHruid();
+        this.email = user.getEmail().orElseThrow(() -> new IllegalStateException("participant must have a valid email"));
```
Comment (Trimax) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/json/UserCreationResponse.java:37]: We're saying **valid** here, but we don't check the e-mail to be valid. Just not null. Isn't it a bit misleading?

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/AdminLookupInvitationRoute.java [hunk 17]
```diff
@@ -63,12 +63,16 @@ public Object handle(Request request, Response response, LookupInvitationPayload
                 }
 
                 String loginEmail = null;
-                if (user.getAuth0UserId() != null) {
+                
+                if (user.getAuth0UserId().isPresent()) {
```
Comment (Trimax) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/AdminLookupInvitationRoute.java]: Regarding returning optionals.
Not sure, but if we'll define the method `hasAuth0UserId` the code will look like
```
if (user.hasAuth0UserId()) {
  ... = user.getAuth0UserId(); //without additional .get statement.
}
```

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/CreateTemporaryUserRoute.java (hunk 4 above)
Comment (Trimax) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/CreateTemporaryUserRoute.java:83]: I was thinking about it. And since we're not using custom builder anymore, I guess the answer is no. But on the other side, if we compare the huge builder class that can be replaced with the one annotation and some places with two calls, I would vote for the 2nd option.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java (hunk 10 above)
Comment (Trimax) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java]: :)
Looks like the work here isn't finished yet.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+            // going to be recognized but handle things just in case. Save the ID since it'll be
+            // needed below to set the `created_by_client_id` field in the user.
+            final var internalClientId = handle.attach(JdbiClient.class)
+                    .getClientIdByAuth0ClientAndDomain(requestorClientId, domain);
+            if (internalClientId.isEmpty()) {
+                /* Disabled for testing!
+                var error = new ApiError(ErrorCodes.NOT_FOUND,
+                        String.format("Auth0 client '%s' is not authorized for '%s'.",
+                            requestorClientId, domain));
+                throw ResponseUtil.haltError(HttpStatus.SC_UNAUTHORIZED, error);
```
Comment (Trimax) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java:81]: Should this be uncommented?

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/UserDeleteService.java [hunk 19]
```diff
@@ -368,12 +368,16 @@ private void deleteElasticSearchData(Handle handle, User user, UserCollectedData
     }
 
     private void deleteAuth0User(User user, Auth0ManagementClient auth0ManagementClient) {
-        if (user.getAuth0UserId() != null && auth0ManagementClient != null) {
-            log.info(LOG_MESSAGE_PREFIX__DELETE_FROM_AUTH, user.getGuid());
-            var result = auth0ManagementClient.deleteAuth0User(user.getAuth0UserId());
-            if (result.hasFailure()) {
-                throw new DDPException(result.hasThrown() ? result.getThrown() : result.getError());
-            }
+        if (user.getAuth0UserId().isPresent() == false) {
```
Comment (Trimax) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/UserDeleteService.java]: Let's better use isEmpty instead?
`if (user.getAuth0UserId().isEmpty())`

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/participants/ParticipantsCreateService.java [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+        /* This is from UserRegistrationRoute.java,
+         * but in that class I can't clearly see where `taskPublisher`
+         * is being set. Thoughts?
+         */
+        /*
+        taskPublisher.publishTask(
+                    TaskPubSubPublisher.TASK_PARTICIPANT_REGISTERED,
+                    payload, studyGuid, participantGuid);
+        */
+    }
```
Comment (Trimax) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/participants/ParticipantsCreateService.java:181]: I don't know as well. Shouldn't it be passed as the constructor argument?

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/studies/StudiesService.java [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+            // Let the caller know.
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    private boolean studyExists(String studyGuid) {
+        var studyId = handle.attach(JdbiUmbrellaStudy.class).getIdByGuid(studyGuid);
+        if (studyId.isPresent()) {
```
Comment (Trimax) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/studies/StudiesService.java]: Maybe just
`return studyId.isPresent()`

Or even
`return handle.attach(JdbiUmbrellaStudy.class).getIdByGuid(studyGuid).isPresent();`

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/util/JavaHeapDumper.java [hunk 22]
```diff
@@ -1,60 +1,161 @@
 package org.broadinstitute.ddp.util;
```
Comment (Trimax) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/util/JavaHeapDumper.java:1]: Is this class expected to be in this PR?
Or do you like my way of concomitant unrelated changes? :)

File: pepper-apis/dss-database/src/main/resources/changelog-master.xml (hunk 5 above)
Comment (Trimax) [reply pepper-apis/dss-database/src/main/resources/changelog-master.xml:5]: Looks like I just accepted IntelliJ's suggestion

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java [hunk 23]
```diff
@@ -334,9 +335,7 @@ private static void start() throws MalformedURLException {
 
         before("*", new HttpHeaderMDCFilter(X_FORWARDED_FOR));
         before("*", new MDCLogBreadCrumbFilter());
-        before("*", (Request request, Response response) -> {
-            MDC.put(MDC_STUDY, RouteUtil.parseStudyGuid(request.pathInfo()));
-        });
+        before("*", (Request request, Response response) -> MDC.put(MDC_STUDY, RouteUtil.parseStudyGuid(request.pathInfo())));
```
Comment (MocanaAtBroad) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java:338]: Not saying this is not a good idea. It is! But dropped in the middle of a 72 file PR that has a completely different purpose without a comment is not the best way to make this change clear and transparent.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/json/UserCreationResponse.java (hunk 16 above)
Comment (bskinner) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/json/UserCreationResponse.java:37]: Nope. Valid in this case is "non-null". I'd be open to changing the text of the exception to make it clearer, but it is not this object's responsibility to validate a User's email. If a User was created with an invalid email, that's a bug elsewhere.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/CreateTemporaryUserRoute.java (hunk 4 above)
Comment (bskinner) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/CreateTemporaryUserRoute.java:83]: That would be fine if these two value were not related. Being able to set both allows the profile to be put into an invalid state.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java (hunk 18 above)
Comment (bskinner) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java:81]: See [this comment](https://github.com/broadinstitute/ddp-study-server/pull/1653/files#r902677146)

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/UserDeleteService.java (hunk 19 above)
Comment (bskinner) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/UserDeleteService.java]: Yup! That's far more straightforward- fixed.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java (hunk 10 above)
Comment (MocanaAtBroad) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java]: I would say we should trust inside of the route that authentication info is A-OK. This work happening in filters/separate concern.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/participants/ParticipantsCreateService.java (hunk 20 above)
Comment (bskinner) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/participants/ParticipantsCreateService.java:181]: Either that or delegated out- either way I don't immediately see where the object is being created (hence my comment), and I don't know what it's lifecycle looks like.

**Review by MocanaAtBroad — COMMENTED** [review]
Still reviewing, but want to submit the comment requesting we remove authentication checks from here. Authorization? I am less sure of that/hope can also be handled in filters.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/util/JavaHeapDumper.java (hunk 22 above)
Comment (bskinner) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/util/JavaHeapDumper.java:1]: See [this comment](https://github.com/broadinstitute/ddp-study-server/pull/1653/files#r902673172)

File: pepper-apis/dss-database/src/main/resources/changelog-master.xml (hunk 5 above)
Comment (bskinner) [reply pepper-apis/dss-database/src/main/resources/changelog-master.xml:5]: Did some digging into this and the other values are technically opaque identifiers, and not URLs- so this is correct (thanks, XML).

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/studies/StudiesService.java (hunk 21 above)
Comment (bskinner) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/studies/StudiesService.java]: I'd go with `return studyId.isPresent();` (mostly because it's easier to step through with a debugger). This whole block is the result of a logical compartmentalization between the Optional check, and the method's returned value.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java (hunk 10 above)
Comment (bskinner) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java]: Yup- see [this comment](https://github.com/broadinstitute/ddp-study-server/pull/1653/files#r902675285)

File: pepper-apis/dss-database/src/main/resources/db-changes/schema/DDP-7931.xml (hunk 3 above)
Comment (MocanaAtBroad) [reply pepper-apis/dss-database/src/main/resources/db-changes/schema/DDP-7931.xml:26]: Maybe this should be "study_center" instead of "center_study"?

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java (hunk 23 above)
Comment (bskinner) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java:338]: Agreed- this was a change in the PR prior to my work.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/db/dao/UserProfileSql.java (hunk 14 above)
Comment (bskinner) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/db/dao/UserProfileSql.java]: Nice catch! Thank you!

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java (hunk 10 above)
Comment (bskinner) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java]: The plan is to remove a lot of this code once I figure out how we're processing the DSM tokens. From what @ssettipalli is telling me, we aren't creating any DDPAuth objects so this code may need to fall-back to manually decoding the token.

**Comment by bskinner** [comment]
Everyone, please leave the "Update branch" button alone.

**Review by MocanaAtBroad — COMMENTED** [review]
Still reviewing, but want to submit the comment requesting we remove authentication checks from here. Authorization? I am less sure of that/hope can also be handled in filters.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+            auth = new DDPAuth("https://domain.example", "abcTESTCLIENTdef", "OPERATOR", "totallyAValidJWT", null, "en");
+        }
+
+        /*
+        * Grab the requesting client id out of the token
+        * so it can be referred back to after we create the user
+        */ 
+        final var requestorClientId = auth.getClient();
+        final var operatorId = auth.getOperator();
+        final var domain = auth.getIssuer();
```
Comment (MocanaAtBroad) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java:44]: I see this property is being picked up from token. Do we know if we get this from DSM tokens @ssettipalli ?

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java [hunk 25]
```diff
@@ -0,0 +1,168 @@
...
+        // Could this be replaced with an "isDSM" check?
+        if (false && auth.hasAdminAccessToStudy(studyGuid) == false) {
+            var error = new ApiError(ErrorCodes.INSUFFICIENT_PRIVILEGES,
+                    String.format("User %s does not have sufficient priviliges for %s.",
+                    operatorId, studyGuid));
+            throw ResponseUtil.haltError(HttpStatus.SC_FORBIDDEN, error);
+        }
+
+        final var email = payload.getEmail();
+        final var emailValidator = EmailValidator.getInstance();
```
Comment (MocanaAtBroad) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java:48]: I believe that if you just use the @Email annotation on the payload POJO, email will be checked for syntax and automatically rejected by ValidatedJsonInputRoute, so we probably don't need to use EmailValidator explicitly.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java [hunk 26]
```diff
@@ -0,0 +1,168 @@
...
+        *
+        * Checking here is in order to save a few cycles & connections to the DB if it's
+        *  immediately clear the client gave us bad data.
+        */
+        if (emailValidator.isValid(email) == false) {
+            var error = new ApiError(ErrorCodes.MALFORMED_EMAIL, "the email address is not in a valid format.");
+            throw ResponseUtil.haltError(HttpStatus.SC_UNPROCESSABLE_ENTITY, error);
+        }
+
+        if (StringUtils.isNotBlank(payload.getCenterId())) {
```
Comment (MocanaAtBroad) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java:63]: Same sort of validation here. You probably can use @NotEmpty (Part of the purpose of ValidatedJsonInputRoute is to check validation annotations).

https://javadoc.io/doc/javax.validation/validation-api/latest/index.html

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/participants/ParticipantsCreateService.java [hunk 27]
```diff
@@ -0,0 +1,182 @@
...
+
+        Long centerId = null;
+
+        if (StringUtils.isNotBlank(centerGuid)) {
+            /* The "centerGuid" is currently just the internal ID
+             * which is a `long` type so convert this to a long before
+             * checking the database. If/when Centers get an separate external
+             * identifier (such as a GUID), then this will need to be updated.
+             */
+            Long tempCenterGuid = Long.valueOf(centerGuid);
```
Comment (MocanaAtBroad) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/participants/ParticipantsCreateService.java:112]: Let's not do this (will not matter for demo as we are not using centerId) but let's use codes/guids whatherver for centers/avoid

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/participants/ParticipantsCreateService.java [hunk 28]
```diff
@@ -0,0 +1,182 @@
...
+    public User createWithAuth0(String userId, String tenantDomain) throws ParticipantCreateError {
+        throw new ParticipantCreateError(Code.NOT_IMPLEMENTED,
+                "Auth0 participant account creation not implemented");
+    }
+
+    private void notifyParticipantWasRegistered(String studyGuid, String operatorGuid, String participantGuid) {
+        var studyDto = handle.attach(JdbiUmbrellaStudy.class).findByStudyGuid(studyGuid);
+        var userDao = handle.attach(JdbiUser.class);
+        var operator = userDao.findByUserGuid(operatorGuid);
+        var participant = userDao.findByUserGuid(participantGuid);
```
Comment (MocanaAtBroad) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/participants/ParticipantsCreateService.java:160]: We can use the  of JdbiUmbrellaStudyCached here.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/studies/StudiesService.java [hunk 29]
```diff
@@ -0,0 +1,120 @@
...
+            // If any id has been returned, the participant is registered
+            // Let the caller know.
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    private boolean studyExists(String studyGuid) {
+        var studyId = handle.attach(JdbiUmbrellaStudy.class).getIdByGuid(studyGuid);
```
Comment (MocanaAtBroad) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/studies/StudiesService.java:117]: JdbiUmbrellaStudyCached. Look in code for how it is instantiated (handle.attach is not how you instantiate Cached Jdbi classes)

**Review by MocanaAtBroad — COMMENTED** [review]
Not seeing test code for the UserCreationRoute.

**Review by MocanaAtBroad — DISMISSED** [review]

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/AdminLookupInvitationRoute.java (hunk 17 above)
Comment (bskinner) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/AdminLookupInvitationRoute.java]: Didn't notice that method! Did a bit of refactoring.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java (hunk 26 above)
Comment (bskinner) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java:63]: This isn't validation  of the input- `centerId` being `null` is perfectly ok, we just don't handle the the case where it is not null yet.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java (hunk 10 above)
Comment (bskinner) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java]: @MocanaAtBroad Code removed

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/json/UserCreationPayload.java (hunk 7 above)
Comment (bskinner) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/json/UserCreationPayload.java]: Spoke with @SimoneMaiwald- this will be transitioned to an opaque string guid when Centers are implemented (like what we're doing in other objects).

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/db/dto/UserDto.java (hunk 15 above)
Comment (bskinner) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/db/dto/UserDto.java:50]: The reasoning for the `Optional<>` here was several-fold:
  * Make it explicit that this was a value that may not be present, and force callers to deal with it (instead of using implicit null handling)
  * Identify places where the method was called, in conjunction with the IDE's "Find References" functionality

While I'd prefer to have this be an Optional, I do agree that this is awkward to use given the current style of our codebase (`model.getEmail().orElse(null);` is also an anti-pattern which should be avoided).

After this latest milestone is reached- let's have a discussion in the process meeting, and figure out what style guidelines we'd like to use in relation to Optionals.

**Review by ssettipalli — APPROVED** [review]

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java (hunk 23 above)
Comment (Trimax) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java:338]: It wasn't dropped, it was refactored. I just removed braces
